### DISCUSSION TITLE
JSAddon

### DISCUSSION THREAD

**Discussion by KhandamovA** [TurboWarp/extensions#809] [Show and tell]
Hello everyone, I want to show you my first extension. This extension adds support for JavaScript functions (I understand that this will most likely not make it into the gallery).

![image](https://github.com/TurboWarp/extensions/assets/77474720/63ecd840-39fc-47a6-9519-b687c21c2566)

executable Js code can return function results directly to "Scratch" environment variables.
Also, the executable js code supports getting values ​​from "Scratch" environment variables if you declare the variables in a special way beforehand.

![image](https://github.com/TurboWarp/extensions/assets/77474720/1a896a2d-6e56-41c9-abec-a9ef59bccdec)
![image](https://github.com/TurboWarp/extensions/assets/77474720/f9d99e94-5f65-4153-9e33-cfd391f23d05)

Also my extension introduces a new way to interact with json objects.

![image](https://github.com/TurboWarp/extensions/assets/77474720/b6bf1d51-84bd-41ae-acfc-3c837d0bb90a)

After creating a JSON variable, we can get it in the variable list.

![image](https://github.com/TurboWarp/extensions/assets/77474720/820770b6-de9f-483b-9914-60e0f4ac70fb)

The main feature is dynamic field access.

![image](https://github.com/TurboWarp/extensions/assets/77474720/9c357af1-7d86-4c17-97f2-be69004946c7)
![image](https://github.com/TurboWarp/extensions/assets/77474720/d017be4d-3d9f-4bf7-8371-f1c743a5fbce)
![image](https://github.com/TurboWarp/extensions/assets/77474720/7b94d180-77c9-429a-beab-fdd23e2d33ff)

There is a special block to get the value

![image](https://github.com/TurboWarp/extensions/assets/77474720/ad194524-2769-40fa-aefc-9ba198f4bb6a)
![image](https://github.com/TurboWarp/extensions/assets/77474720/22e6f758-8cfe-4cf1-88bc-9a805c887ee6)
![image](https://github.com/TurboWarp/extensions/assets/77474720/249d6671-917f-4c46-97ae-fa5005cba58c)
![image](https://github.com/TurboWarp/extensions/assets/77474720/3219634d-ac7b-4662-88f7-4c7912ab8a53)

You can also add and remove keys.

![image](https://github.com/TurboWarp/extensions/assets/77474720/9bff9f78-7a26-48d2-833f-dac864efca0b)
![image](https://github.com/TurboWarp/extensions/assets/77474720/dd867674-9443-41e4-ba1c-a61eeda747cb)

Well, here are the blocks for working with arrays in json objects.

![image](https://github.com/TurboWarp/extensions/assets/77474720/b3708cd1-f968-4f98-b0c5-46bae43d099a)

What are your opinions and suggestions?

**Comment by PURPL3-CAT** [comment]
This looks cool! Is there a repo?

**Reply by KhandamovA** [reply]
I will share my extension a bit later, at the moment I am preparing an update of my extension, which will also be useful (I hope so).

**Reply by KhandamovA** [reply]
https://github.com/KhandamovA/useful-js-libs/blob/main/TurboWarp/JSAddon.js
